src/fwupd_test.py

```diff
@@ -0,0 +1,345 @@
+                "org.freedesktop.fwupd.modify-remote",
+                "org.freedesktop.fwupd.set-approved-firmware",
+                "org.freedesktop.fwupd.self-sign",
+                "org.freedesktop.fwupd.get-bios-settings",
+                "org.freedesktop.fwupd.set-bios-settings",
+                "org.freedesktop.fwupd.fix-host-security-attr",
+                "org.freedesktop.fwupd.undo-host-security-attr",
+            ]
+        )
+
+        self.proxy = None
+        self.props_proxy = None
+        self.daemon_log = None
+        self.daemon = None
+        self.changed_properties = {}
+
+    def run(self, result=None):
+        super().run(result)
+        if not result or not self.daemon_log:
+            return
+        if len(result.errors) + len(result.failures) or os.getenv("VERBOSE"):
+            with open(self.daemon_log.name, encoding="utf-8") as tmpf:
+                sys.stderr.write("\nDAEMON log:\n")
+                sys.stderr.write(tmpf.read())
+                sys.stderr.write("\n")
+
+    def tearDown(self):
+        del self.testbed
+        self.stop_daemon()
+
+        if self.polkitd:
+            self.polkitd.stdout.close()
+            try:
+                self.polkitd.kill()
+            except OSError:
+                pass
+            self.polkitd.wait()
+
+        self.obj_polkit = None
+
+    #
+    # Daemon control and D-BUS I/O
+    #
+
+    def start_daemon(self):
+        """Start daemon and create DBus proxy.
+
+        When done, this sets self.proxy as the Gio.DBusProxy for power-profiles-daemon.
+        """
+        env = os.environ.copy()
+        env["G_DEBUG"] = "fatal-criticals"
+        env["G_MESSAGES_DEBUG"] = "all"
+        # note: Python doesn't propagate the setenv from Testbed.new(), so we
+        # have to do that ourselves
+        env["UMOCKDEV_DIR"] = self.testbed.get_root_dir()
+        self.daemon_log = (
+            tempfile.NamedTemporaryFile()
+        )  # pylint: disable=consider-using-with
+        daemon_path = [self.daemon_path, "-vv"]
+
+        # pylint: disable=consider-using-with
+        self.daemon = subprocess.Popen(
+            daemon_path, env=env, stdout=self.daemon_log, stderr=subprocess.STDOUT
+        )
+        self.addCleanup(self.daemon.kill)
+
+        def on_proxy_connected(_, res):
+            try:
+                self.proxy = Gio.DBusProxy.new_finish(res)
+            except GLib.Error as exc:
+                self.fail(exc)
+
+        cancellable = Gio.Cancellable()
+        self.addCleanup(cancellable.cancel)
+        Gio.DBusProxy.new(
+            self.dbus,
+            Gio.DBusProxyFlags.DO_NOT_AUTO_START,
+            None,
+            self.DBUS_NAME,
+            self.DBUS_PATH,
+            self.DBUS_INTERFACE,
+            cancellable,
+            on_proxy_connected,
+        )
+
+        # wait until the daemon gets online
+        wait_time = 20 if "valgrind" in daemon_path[0] else 5
```

Comment:
i think bump these a bit -- I'm thinking things like s390 emulation?

Reply:
20 seconds isn't enough?

Reply:
nah, maybe make the valgrind one 60 seconds...